## Or Preprocessing + Download

In this case we use a custom made function that applies a preprocessing after downloading each files in the database.  
在这种情况下，我们使用一个定制的函数，在下载数据库中的每个文件后应用预处理。

In [97]:
#用于数据预处理的类
from scdataloader.preprocess import (
    LaminPreprocessor,
    additional_postprocess,
    additional_preprocess,
)

In [6]:
#定义描述信息
DESCRIPTION='preprocessed by scDataLoader'

In [7]:
# Here we also add some additional preprocessing (happens at the beginning of the preprocessing function) and post processing (happens at the end of the preprocessing function)
# 这里我们还添加了一些额外的预处理（发生在预处理函数的开始）和后处理（发生在预处理函数的结束）。
# this serves as an exemple of the flexibility of the function
# 这是函数灵活性的一个例子 就是自己定义做什么 
# 初始化 LaminPreprocessor
do_preprocess = LaminPreprocessor(additional_postprocess=additional_postprocess, additional_preprocess=additional_preprocess, skip_validate=True, subset_hvg=0)


In [16]:
preprocessed_dataset = do_preprocess(cx_dataset, name=DESCRIPTION, description=DESCRIPTION, start_at=6, version="2")
#输出Artifact 对象生成一个预处理后的 AnnData 对象

0
Artifact(uid='Mgilie8RUip2slElQoDx', key='cell-census/2023-12-15/h5ads/77044335-0ac5-4406-9b3f-8cdd3656d67b.h5ad', suffix='.h5ad', accessor='AnnData', description='Dissection: Pons (Pn) - afferent nuclei of cranial nerves in pons - PnAN', version='2023-12-15', size=131533988, hash='QCaNEZY9apeRmazmwGXAWg-16', hash_type='md5-n', n_observations=23349, visibility=1, key_is_virtual=False, updated_at=2024-01-29 07:45:42 UTC, storage_id=2, transform_id=16, run_id=22, created_by_id=1)
AnnData object with n_obs × n_vars = 23349 × 59357
    obs: 'roi', 'organism_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'assay_ontology_term_id', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'donor_id', 'suspension_type', 'dissection', 'fraction_mitochondrial', 'fraction_unspliced', 'cell_cycle_score', 'total_genes', 'total_UMIs', 'sample_id', 'supercluster_term', 'cluster_id', 'subcluster_id', 'cell_type_ontology_term_id', 'tissue_ontology_term_

/home/ml4ig1/miniconda3/envs/test/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
  warnings.warn(


❗ no run & transform get linked, consider passing a `run` or calling ln.track()


... storing 'dpt_group' as categorical
... storing 'symbol' as categorical
... storing 'ncbi_gene_ids' as categorical
... storing 'biotype' as categorical
... storing 'description' as categorical
... storing 'synonyms' as categorical
... storing 'organism' as categorical


1
Artifact(uid='iAZPSOBKLpaK7lqyYp9O', key='cell-census/2023-12-15/h5ads/2a8ca8f3-5599-4cda-b973-3a2dfc3c1fe6.h5ad', suffix='.h5ad', accessor='AnnData', description='Dissection: Amygdaloid complex (AMY) - Corticomedial nuclear group (CMN) - anterior cortical nucleus - CoA', version='2023-12-15', size=169529860, hash='WuShpnxfduKWKn23oYUCTA-21', hash_type='md5-n', n_observations=10778, visibility=1, key_is_virtual=False, updated_at=2024-01-29 07:45:42 UTC, storage_id=2, transform_id=16, run_id=22, created_by_id=1)
... downloading 2a8ca8f3-5599-4cda-b973-3a2dfc3c1fe6.h5ad: 100.0%
AnnData object with n_obs × n_vars = 10778 × 59357
    obs: 'roi', 'organism_ontology_term_id', 'disease_ontology_term_id', 'self_reported_ethnicity_ontology_term_id', 'assay_ontology_term_id', 'sex_ontology_term_id', 'development_stage_ontology_term_id', 'donor_id', 'suspension_type', 'dissection', 'fraction_mitochondrial', 'fraction_unspliced', 'cell_cycle_score', 'total_genes', 'total_UMIs', 'sample_id', 'sup

/home/ml4ig1/miniconda3/envs/test/lib/python3.10/site-packages/anndata/_core/anndata.py:522: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.
  warnings.warn(


KeyboardInterrupt: 

#### Artifact 对象和AnnData 对象关系  
1. 封装与描述：  
Artifact 对象封装了一个具体的文件（如 .h5ad 文件），并提供了对该文件的描述信息和元数据。  
Artifact 对象的 accessor 字段指定了如何访问该文件中的数据，例如通过 AnnData 访问器读取 .h5ad 文件中的 AnnData 对象。  
2. 数据存储与访问：  
Artifact 对象描述了文件的位置、版本、大小等元数据，并提供了一种标准化的方式来访问和管理这些文件。
AnnData 对象则是实际存储在这些文件中的数据结构，包含了单细胞数据的各种维度和元数据。
3. 实例示例：  
在你的输出中，Artifact 对象描述了一个名为 'cell-census/2023-12-15/h5ads/77044335-0ac5-4406-9b3f-8cdd3656d67b.h5ad' 的 .h5ad 文件，其 accessor 为 'AnnData'，表示可以通过 AnnData 访问器读取该文件中的数据。  
接下来，AnnData 对象展示了从该文件加载的实际数据，包括观测数 (n_obs = 23349) 和变量数 (n_vars = 59357)，以及相关的元数据字段。  

In [17]:
#we have processed that many files 统计符合处理过的文件数量
len(ln.Artifact.filter(version='2', description=DESCRIPTION))

1

In [18]:
# we can load a preprocessed anndata like this 从一个 Artifact 对象中加载预处理后的 AnnData 数据，并访问其观测数据（obs）。
adata = ln.Artifact.filter(version='2', description=DESCRIPTION)[0].backed()
adata.obs

,roi,organism_ontology_term_id,disease_ontology_term_id,self_reported_ethnicity_ontology_term_id,assay_ontology_term_id,sex_ontology_term_id,development_stage_ontology_term_id,donor_id,suspension_type,dissection,...,total_counts_hb,log1p_total_counts_hb,pct_counts_hb,outlier,mt_outlier,leiden_3,leiden_2,leiden_1,dpt_group,heat_diff
73b14343-9071-4da1-884d-52f04c781a44,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000136,H19.30.001,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,3.0,1.386294,0.023204,True,False,13,11,10,,NaN
91e3096c-9a19-4827-8050-9e08fe22f9e0,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000136,H19.30.001,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,0.0,0.000000,0.000000,False,False,5,4,3,,NaN
7bd7c1f8-5080-4ea4-87ff-4c02bf641059,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000136,H19.30.001,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,2.0,1.098612,0.016004,True,False,19,17,6,,NaN
47d8e4b5-daf3-4b3f-bdaa-c877fce88f08,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000144,H18.30.002,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,3.0,1.386294,0.008974,True,True,24,20,16,,NaN
5bc7626d-4e45-4bf1-8818-922502760364,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000144,H18.30.002,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,0.0,0.000000,0.000000,True,False,34,27,22,,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
f266c95f-e602-416b-9f3e-c8f081352e79,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000144,H18.30.002,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,2.0,1.098612,0.032889,False,False,12,9,9,9_PATO:0000461_CL:0002453_UBERON:0000988,0.048092
02530a0a-bd8b-40ff-b3ab-18bbbac42eda,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000144,H18.30.002,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,1.0,0.693147,0.011792,False,False,12,9,9,9_PATO:0000461_CL:0002453_UBERON:0000988,0.049665
a1840018-61f1-4345-b665-90ce5ad7936d,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000144,H18.30.002,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,0.0,0.000000,0.000000,False,False,12,9,9,9_PATO:0000461_CL:0002453_UBERON:0000988,0.049922
02c6c139-4346-443b-8209-9ab51d80595c,Human PnAN,NCBITaxon:9606,PATO:0000461,HANCESTRO:0005,EFO:0009922,PATO:0000384,HsapDv:0000144,H18.30.002,nucleus,Pons (Pn) - afferent nuclei of cranial nerves ...,...,0.0,0.000000,0.000000,False,False,12,9,9,9_PATO:0000461_CL:0002453_UBERON:0000988,0.056045


ln.Artifact.filter(version='2', description=DESCRIPTION):
1. 这里调用了 ln.Artifact 类的 .filter() 方法，用于筛选符合特定条件的工件（文件）。
version='2': 筛选版本号为 '2' 的工件。  
description=DESCRIPTION: 筛选描述信息等于 DESCRIPTION 变量值的工件。  
2. [0]:
使用索引 [0] 获取筛选结果中的第一个工件。假设 .filter() 返回的是一个列表或类似列表的对象，这里取第一个元素。  
3. .backed():  
调用 .backed() 方法，表示以“后备”模式加载 AnnData 对象。在后备模式下，数据集不会完全加载到内存中，而是保持与磁盘上的文件同步。这对于处理非常大的数据集特别有用，因为它可以节省内存。

In [20]:
# I need to remake the dataset as it failed for some files and I had to restart at position 11 (As you can see in the preprocess() function)
name="preprocessed dataset"
description="preprocessed dataset using scdataloader"
dataset = ln.Collection(ln.Artifact.filter(version='2', description=DESCRIPTION), name=name, description=description)
dataset.save()
dataset.artifacts.count()

1

#### ln.Collection：
这是一个类或函数，用于创建一个新的集合对象，通常用于管理和组织多个工件（Artifact）。  
参数：  
ln.Artifact.filter(version='2', description=DESCRIPTION)：筛选出符合条件的工件（文件），这里是版本号为 '2' 且描述信息为 DESCRIPTION 的工件。 
name：数据集的名称。  
description：数据集的描述信息。 

dataset.save()：
将新创建的数据集保存到持久化存储中。这一步确保数据集及其元数据被正确地存储，以便后续使用。  

dataset.artifacts.count()：
获取数据集中包含的工件（Artifact）的数量。这可以帮助你确认数据集是否正确地包含了所有预期的工件。  

**dataset 并不是一个 Artifact 对象，而是一个 Collection 对象。Collection 对象用于管理和组织多个 Artifact 对象。下面详细解释这两者的区别及其关系。**

个人感想：这个教程感觉还是需要和lamin一起用 我能不能像我一开始想的那样直接处理GSE数据集？  我感觉还是得回到他的代码  看看有没有别的函数 可以让AI帮我解释学习一下

## 尝试用一个Preprocessor实例来处理我的h5ad文件

In [1]:
from scdataloader.preprocess import Preprocessor
import anndata as ad
import scanpy as sc

→ connected lamindb: zhang0730/lamin_data


In [9]:
#创建 Preprocessor 实例：根据你的需求设置 Preprocessor 类的初始化参数。例如：
preprocessor = Preprocessor(
    filter_gene_by_counts=5,  # 可以根据实际情况调整参数
    filter_cell_by_counts=5,
    normalize_sum=1e4,
    n_hvg_for_postp=2000,
    hvg_flavor="seurat",
    min_valid_genes_id=1000,  # 用于指定一个数据集在处理过程中需要保留的最少有效基因数量。
    binning=3,
    organisms = ["NCBITaxon:9606"],
    is_symbol = True
    # 其他参数根据需要设置
)

In [3]:
adata = ad.read_h5ad("/disk1/cai026/Mellon/data/preprocessed_t-cell-depleted-bm-rna.h5ad")
#不能是鼠的 只能是人的

In [4]:
adata.obs

,sample,n_genes_by_counts,total_counts,total_counts_mt,pct_counts_mt,batch,DoubletScores,n_counts,leiden,phenograph,log_n_counts,celltype,palantir_pseudotime,selection,NaiveB_lineage,mellon_log_density,mellon_log_density_clipped
IM-1393_BoneMarrow_TcellDep_1_multiome#GTGAGCGAGTCTCACC-1,IM-1393_BoneMarrow_TcellDep_1_multiome,2501,5612.0,735.0,13.096934,0,0.061947,5593.0,7,3,3.749118,MonoPre,0.138621,False,False,101.548018,101.548018
IM-1393_BoneMarrow_TcellDep_1_multiome#GAGTCAAAGTCCTTCA-1,IM-1393_BoneMarrow_TcellDep_1_multiome,5015,21563.0,1883.0,8.732552,0,0.050938,21505.0,8,9,4.333709,EryPre1,0.336803,False,False,115.809166,115.809166
IM-1393_BoneMarrow_TcellDep_1_multiome#TGTGCGCAGTCGCTAG-1,IM-1393_BoneMarrow_TcellDep_1_multiome,1102,2297.0,50.0,2.176752,0,0.049296,2276.0,13,12,3.361161,Ery,0.684445,False,False,107.371474,107.371474
IM-1393_BoneMarrow_TcellDep_1_multiome#ATATGTCCAATGCCTA-1,IM-1393_BoneMarrow_TcellDep_1_multiome,2575,5653.0,663.0,11.728286,0,0.106040,5616.0,3,0,3.752279,Mono,0.311772,False,False,177.693448,177.693448
IM-1393_BoneMarrow_TcellDep_1_multiome#CTTAGTTTCGCTAGTG-1,IM-1393_BoneMarrow_TcellDep_1_multiome,1262,2651.0,68.0,2.565070,0,0.059961,2618.0,13,12,3.423410,Ery,0.704443,False,False,102.709978,102.709978
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
IM-1393_BoneMarrow_TcellDep_2_multiome#ATCCGTGAGGGATTAG-1,IM-1393_BoneMarrow_TcellDep_2_multiome,1968,3902.0,393.0,10.071758,1,0.037037,3895.0,0,0,3.591287,Mono,0.309793,False,False,159.251494,159.251494
IM-1393_BoneMarrow_TcellDep_2_multiome#ATTTAGGTCAGGTTTA-1,IM-1393_BoneMarrow_TcellDep_2_multiome,2264,4902.0,413.0,8.425133,1,0.050847,4895.0,2,1,3.690373,Mono,0.306938,False,False,165.952314,165.952314
IM-1393_BoneMarrow_TcellDep_2_multiome#GATGGACAGATAAAGC-1,IM-1393_BoneMarrow_TcellDep_2_multiome,2371,4947.0,286.0,5.781282,1,0.129032,4931.0,1,0,3.694342,Mono,0.311873,False,False,179.788002,179.788002
IM-1393_BoneMarrow_TcellDep_2_multiome#GAAGGCTAGCTATATG-1,IM-1393_BoneMarrow_TcellDep_2_multiome,1764,3554.0,440.0,12.380417,1,0.033175,3545.0,0,0,3.550717,Mono,0.312016,False,False,179.480591,179.480591


需要有'organism_ontology_term_id'列  但我看他的参考数据的这种数据一看就是有专门的东西生成的呀 我直接自己生成

In [5]:
adata.obs['organism_ontology_term_id'] = "NCBITaxon:9606" #这是人类的
adata.obs['self_reported_ethnicity_ontology_term_id'] = "unknown"
adata.obs['sex_ontology_term_id'] = "unknown"
adata.obs['disease_ontology_term_id'] = "healthy"
#https://disease-ontology.org/do/
adata.obs['cell_type_ontology_term_id'] = "unknown"
adata.obs['development_stage_ontology_term_id'] = "unknown"

In [6]:
#https://www.ebi.ac.uk/ols/
adata.obs['tissue_ontology_term_id'] = "UBERON:0002371"
#https://www.ebi.ac.uk/ols4/ontologies/uberon/classes/http%253A%252F%252Fpurl.obolibrary.org%252Fobo%252FUBERON_0002371

In [7]:
#https://www.ebi.ac.uk/ols4/ontologies/efo/classes/http%253A%252F%252Fwww.ebi.ac.uk%252Fefo%252FEFO_0008913
adata.obs['assay_ontology_term_id'] = "EFO:0008913"

## 自动注释数据集中的 cell_type_ontology_term_id 列  
通常需要将现有的 celltype 列映射到相应的本体术语（Ontology Terms）。

### 1. 准备映射表

In [77]:
# 示例映射表
celltype_to_ontology = {
    "T cell": "CL:0000084",
    "B cell": "CL:0000236",
    "Macrophage": "CL:0000235",
    "Neuron": "CL:0000105",
    # 添加更多的映射关系
}

### 2. 查询本体数据库（可选）
如果你没有现成的映射表，可以使用本体数据库（如 bionty 或 cell_ontology）来查询并生成映射表。以下是一个使用 bionty 库的示例：

In [78]:
import bionty as bt

def get_cell_ontology_term(celltype):
    try:
        term = bt.CellType.filter(name=celltype).first()
        return term.ontology_id
    except Exception as e:
        print(f"Error fetching ontology term for {celltype}: {e}")
        return "unknown"

# 示例 celltype 列
celltypes = ["T cell", "B cell", "Macrophage", "Neuron"]

# 获取映射表
celltype_to_ontology = {celltype: get_cell_ontology_term(celltype) for celltype in celltypes}
print(celltype_to_ontology)

Error fetching ontology term for Macrophage: 'NoneType' object has no attribute 'ontology_id'
Error fetching ontology term for Neuron: 'NoneType' object has no attribute 'ontology_id'
{'T cell': 'CL:0000084', 'B cell': 'CL:0000236', 'Macrophage': 'unknown', 'Neuron': 'unknown'}


### 3. 应用映射
将映射表应用到 AnnData 对象的 .obs 表中：

In [ ]:
import scanpy as sc
import pandas as pd

# 假设你已经有一个 AnnData 对象 adata，并且有一列 'celltype'
adata = sc.read_h5ad('your_dataset.h5ad')

# 示例映射表
celltype_to_ontology = {
    "T cell": "CL:0000084",
    "B cell": "CL:0000236",
    "Macrophage": "CL:0000235",
    "Neuron": "CL:0000105",
    # 添加更多的映射关系
}

# 创建一个新的列 'cell_type_ontology_term_id' 并填充映射值
adata.obs['cell_type_ontology_term_id'] = adata.obs['celltype'].map(celltype_to_ontology)

# 检查结果
print(adata.obs[['celltype', 'cell_type_ontology_term_id']].head())

## 自动注释数据集中的 disease_ontology_term_id 列  
通常需要将现有的 disease 列映射到相应的本体术语（Ontology Terms）。

### 1. 准备映射表

In [77]:
# 示例映射表
disease_to_ontology = {
    "Healthy": "PATO:0000461",  # Healthy is often represented as PATO:0000461 (normal)
    "Type 2 Diabetes": "DOID:9352",
    "Hypertension": "DOID:10763",
    "Asthma": "DOID:2841",
    # 添加更多的映射关系
}

### 2. 查询本体数据库（可选）
如果你没有现成的映射表，可以使用本体数据库（如 bionty 或 cell_ontology）来查询并生成映射表。以下是一个使用 bionty 库的示例：

In [ ]:
import bionty as bt

def get_disease_ontology_term(disease):
    try:
        term = bt.Disease.filter(name=disease).first()
        return term.ontology_id
    except Exception as e:
        print(f"Error fetching ontology term for {disease}: {e}")
        return "unknown"

# 示例 disease 列
diseases = ["Healthy", "Type 2 Diabetes", "Hypertension", "Asthma"]

# 获取映射表
disease_to_ontology = {disease: get_disease_ontology_term(disease) for disease in diseases}
print(disease_to_ontology)

### 3. 应用映射
将映射表应用到 AnnData 对象的 .obs 表中：

In [ ]:
import scanpy as sc
import pandas as pd

# 假设你已经有一个 AnnData 对象 adata，并且有一列 'disease'
adata = sc.read_h5ad('your_dataset.h5ad')

# 示例映射表
disease_to_ontology = {
    "Healthy": "PATO:0000461",  # Healthy is often represented as PATO:0000461 (normal)
    "Type 2 Diabetes": "DOID:9352",
    "Hypertension": "DOID:10763",
    "Asthma": "DOID:2841",
    # 添加更多的映射关系
}

# 创建一个新的列 'disease_ontology_term_id' 并填充映射值
adata.obs['disease_ontology_term_id'] = adata.obs['disease'].map(disease_to_ontology)

# 检查结果
print(adata.obs[['disease', 'disease_ontology_term_id']].head())

In [10]:
#调用 Preprocessor 实例处理数据：
processed_adata = preprocessor(adata)

Dropping layers:  KeysView(Layers with keys: )
checking raw counts
removed 0 non primary cells, 8627 renamining
filtered out 0 cells, 8627 renamining
Removed 0 genes.
validating
! received 1 unique term, 8626 empty/duplicated terms are ignored
! 1 unique term (100.00%) is not validated for ontology_id: 'healthy'
! received 1 unique term, 8626 empty/duplicated terms are ignored
! 1 unique term (100.00%) is not validated for ontology_id: 'unknown'
startin QC


/disk1/cai026/mambaforge/envs/model/lib/python3.10/site-packages/scdataloader/preprocess.py:254: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  adata, organism=adata.obs.organism_ontology_term_id[0], need_all=False


Seeing 5422 outliers (62.85% of total dataset):
normalize


/disk1/cai026/mambaforge/envs/model/lib/python3.10/site-packages/scdataloader/preprocess.py:361: ImplicitModificationWarning: Trying to modify index of attribute `.obs` of view, initializing view as actual.
  adata.obs.index = [str(uuid4()) for _ in range(adata.shape[0])]


Binning data ...
non_zero_row shape: (1, 2361)
non_zero_row content: [[1. 1. 2. ... 2. 1. 2.]]


ValueError: shapes (2,2361) and (2,1) not aligned: 2361 (dim 1) != 2 (dim 0)

**Exception: Dataset dropped due to too many genes not mapping to it**  
原因是经过基因筛选后，剩余的基因数量小于 min_valid_genes_id。
 min_valid_genes_id=5000  # 将其调整为一个合适的较小值
检查基因注释的一致性  
**data需要是raw count**  
最后解决方法是找源码234 intersect_genes = set(adata.var.index).intersection(set(genesdf.index)) 觉得可能是ENS—ID设置的问题   里面_init_的属性设置is_symbol = True

In [21]:
adata.var
#No ENS genes found, assuming gene symbols...
#代码预期的基因名格式可能是 Ensembl 基因 ID（以 “ENS” 开头），但你的数据并不符合这个格式，所以被当作基因符号来处理。

,n_cells,highly_variable,means,dispersions,dispersions_norm,PeakCounts,n_counts
AL627309.1,103,False,0.032396,3.376752,0.271279,0,85.0
AL627309.5,387,False,0.125132,3.302549,0.216766,0,334.0
LINC01409,993,False,0.289841,3.125480,-0.800211,0,877.0
LINC01128,627,True,0.184895,3.387681,1.553986,0,556.0
LINC00115,292,False,0.086229,3.322483,0.051071,3,241.0
...,...,...,...,...,...,...,...
MT-ND6,6160,False,2.260383,1.858314,0.659850,0,10166.0
MT-CYB,10966,False,8.243164,0.094619,-3.384264,0,353502.0
MAFIP,413,False,0.107827,3.118896,-1.502928,0,311.0
AL354822.1,185,False,0.045086,3.135284,-1.516408,0,137.0


In [29]:
import scanpy as sc

# 打印初始基因和细胞数量
print(f"Initial number of genes: {adata.shape[1]}")
print(f"Initial number of cells: {adata.shape[0]}")

# 过滤基因
if preprocessor.filter_gene_by_counts is not None:
    sc.pp.filter_genes(adata, min_counts=preprocessor.filter_gene_by_counts)
    print(f"After filtering by gene counts (min_counts={preprocessor.filter_gene_by_counts}): {adata.shape[1]} genes remaining")

# 过滤细胞
if preprocessor.filter_cell_by_counts is not None:
    sc.pp.filter_cells(adata, min_counts=preprocessor.filter_cell_by_counts)
    print(f"After filtering by cell counts (min_counts={preprocessor.filter_cell_by_counts}): {adata.shape[0]} cells remaining")

Initial number of genes: 17220
Initial number of cells: 8627
After filtering by gene counts (min_counts=5): 17220 genes remaining
After filtering by cell counts (min_counts=5): 8627 cells remaining


In [53]:
from scdataloader import utils as data_utils
# 假设'NCBITaxon:9606'是你要查询基因集对应的物种本体术语ID
genes_df = data_utils.load_genes('NCBITaxon:9606')
print(genes_df)

                          uid     symbol ncbi_gene_ids         biotype  \
ensembl_gene_id                                                          
ENSG00000000003  1uTi9dROoaN5     TSPAN6          7105  protein_coding   
ENSG00000000005  2FhduD7Z97Uv       TNMD         64102  protein_coding   
ENSG00000000419  4eC1wUNJAO2s       DPM1          8813  protein_coding   
ENSG00000000457  5Zug63FETk4p      SCYL3         57147  protein_coding   
ENSG00000000460  3wEqb6eOTZzC      FIRRM         55732  protein_coding   
...                       ...        ...           ...             ...   
ENSG00000293596  42lwxhgTHxaN       None     105372654          lncRNA   
ENSG00000293597  12nPu7ce9hLZ  LINC00970     101978719          lncRNA   
ENSG00000293599  2DLKDsGV0zX9       None                        lncRNA   
ENSG00000293600  7hbNrFh6GVRt       None     131768270  protein_coding   
ENSG00000293604  43PKjUN2TuRj      ORAI1         84876  protein_coding   

                                     

In [54]:
import pandas as pd
# adata.var.index 包含基因名
adata.var.index = adata.var.index.str.upper()
genes_df['symbol'] = genes_df['symbol'].str.upper()

In [55]:
adata_genes = set(adata.var.index)
ref_genes = set(genes_df['symbol'])
unmatched_genes = adata_genes - ref_genes
print(unmatched_genes)

{'AC016907.2', 'AC008114.1', 'AC002553.2', 'AC007032.1', 'AC104806.2', 'AC010245.2', 'AL117381.1', 'AC104248.1', 'AC103831.1', 'AL358472.5', 'AC012485.1', 'AC087284.1', 'ZBED6CL', 'AC104134.1', 'AL355581.1', 'AC126283.1', 'AC132153.1', 'AL031595.2', 'AC083798.2', 'AC079841.2', 'STX17-AS1', 'AC017101.1', 'FP236383.3', 'AC099501.1', 'AC244502.1', 'AC007991.3', 'AC067852.3', 'AC087645.2', 'AC025287.4', 'AL645465.1', 'AC015871.4', 'AC073342.1', 'AC097448.1', 'AC108210.1', 'HIST1H2BL', 'AC104335.1', 'AC008906.1', 'AL139241.1', 'FP671120.4', 'HIST1H3A', 'MARCH1', 'AC244090.1', 'AL096794.1', 'AL390195.2', 'H2AFX', 'AP000873.2', 'AC005224.2', 'AC091729.3', 'AC011997.2', 'AL445584.2', 'AC009061.2', 'AC091769.2', 'HIST1H2AK', 'AP001372.2', 'AC012076.1', 'AL359541.1', 'AC005695.2', 'AC245452.1', 'AC068888.1', 'AL670729.2', 'AL589693.1', 'AL136038.5', 'AC005670.3', 'AC084036.1', 'AC023509.6', 'AC234582.1', 'AC008894.2', 'VNN3', 'AC097460.1', 'AL031123.4', 'AL022068.1', 'AC011472.2', 'AC131237.1', 